### PR TITLE
Table group joins support in hql

### PR DESCRIPTION
**Pull request by bahusoid**
Loose port of https://github.com/hibernate/hibernate-orm/commit/753858ce735b2a4eb1a88e3dcd676a0f4e444c2b

Example of SQL generated with "group joins":
```sql
SELECT *
FROM Person person0_
INNER JOIN (
	IndividualCustomer individual1_ 
	INNER JOIN Customer individual1_1_ ON individual1_.IndividualCustomerID = individual1_1_.Id
	) ON person0_.Id = individual1_.PersonID AND individual1_1_.Deleted = @p0
```

It's useful when withClause needs to be applied to multiple tables:
* subclasses with withClause referencing base tables
* Complex withClause with implied joins (like `from a in A join b in B b on a.Id equals b.C.D.Id` also see added test case `WithImpliedJoinOnAssociation`) 
* Many-to-many with filters/ withClause (makes fix implemented in https://github.com/nhibernate/nhibernate-core/pull/1995 obsolete as with group joins no null data is retrieved)

Fixes hql/Linq for #1209 and #1215

### CODE REVIEW

File: src/NHibernate.Test/Hql/Ast/WithClauseFixture.cs [hunk 1]
```diff
@@ -46,36 +44,26 @@ public void ValidWithSemantics()
...
-					s.CreateQuery("from Human h inner join h.offspring o with o.mother.father = :cousin")
-					.SetInt32("cousin", 123)
-					.List();
-				}
-				catch (GenericADOException)
-				{
-					//Apparently SQLite can process queries with wrong join orders
-				}
+					{
+						Assert.DoesNotThrow(
```
Comment (bahusoid) [inline src/NHibernate.Test/Hql/Ast/WithClauseFixture.cs:51]: @maca88 This test generates invalid async code

Comment (maca88) [reply src/NHibernate.Test/Hql/Ast/WithClauseFixture.cs]: Fixed in #2363.

**Comment by bahusoid** [comment]
It's done! It's a good stuff :)

File: src/NHibernate/SqlCommand/ANSIJoinFragment.cs [hunk 2]
```diff
@@ -66,6 +46,33 @@ public override void AddJoin(string tableName, string alias, string[] fkColumns,
 			AddCondition(_fromFragment, on);
 		}
 
+		internal static string GetJoinString(JoinType joinType)
+		{
+			string joinString;
+			switch (joinType)
+			{
+				case JoinType.InnerJoin:
+					joinString = " inner join ";
```
Comment (hazzik) [inline src/NHibernate/SqlCommand/ANSIJoinFragment.cs:55]: can we please `return " inner join "`?

File: src/NHibernate/Hql/Ast/ANTLR/Tree/FromElement.cs [hunk 3]
```diff
@@ -311,8 +315,11 @@ public FromElement RealOrigin
 		public SqlString WithClauseFragment
 		{
 			get { return _withClauseFragment; }
+			set => _withClauseFragment = value;
```
Comment (hazzik) [inline src/NHibernate/Hql/Ast/ANTLR/Tree/FromElement.cs:318]:
```suggestion
			set { _withClauseFragment = value; }
```

**Comment by fredericDelaporte** [comment]
If it is done, can you please remove the WIP in the branch name?

(Free nitpick of course, to be ignored ;)

**Comment by bahusoid** [comment]
>If it is done, can you please remove the WIP in the branch name?

@fredericDelaporte I don't know how to do it without closing this and opening new PR... 😕

**Comment by fredericDelaporte** [comment]
Yes, it is just a joke from me.

**Review by hazzik — APPROVED** [review]

**Pull request merged**